stan/math/prim/mat/fun/mrrr.hpp

```diff
@@ -0,0 +1,709 @@
+        const double max_shift = (high[i] - low[cluster_end]) * 10;
+        double next_shift, min_ele_growth;
+        find_shift(block.l, block.d, low[cluster_end], high[i], max_ele_growth,
+                   max_shift, l, d, next_shift, min_ele_growth);
+        for (int j = i; j <= cluster_end; j++) {
+          low[j] = low[j] * (1 - copysign(shift_error, low[j])) - next_shift;
+          high[j] = high[j] * (1 + copysign(shift_error, high[j])) - next_shift;
+          eigenval_bisect_refine(l, d, low[j], high[j], j);
+        }
+        block_queue.push(mrrr_task{i, cluster_end + 1, block.shift + next_shift,
+                                   std::move(l), std::move(d),
+                                   block.level + 1});
+        l.resize(n - 1);  // after move out
+        d.resize(n);
+
+        i = cluster_end;
+      } else {  // isolated eigenvalue
+        int twist_idx;
+        const double low_gap = i == block.start
+                                   ? std::numeric_limits<double>::infinity()
+                                   : low[i - 1] - high[i];
+        const double high_gap = i == block.end - 1
+                                    ? std::numeric_limits<double>::infinity()
+                                    : low[i] - high[i + 1];
+        const double min_gap = std::min(low_gap, high_gap);
+        const Eigen::VectorXd *l_ptr, *d_ptr;
+        if (!(fabs(min_gap / ((high[i] + low[i]) * 0.5)) > min_rel_sep)) {
+          if (!(fabs(min_gap / ((high[i] + low[i]) * 0.5 - shift)) > min_rel_sep
+                && min_element_growth < max_ele_growth)) {
+            const double max_shift = min_gap / min_rel_sep;
+            find_shift(block.l, block.d, low[i], high[i], max_ele_growth,
+                       max_shift, l2, d2, shift, min_element_growth);
+          }
+          low[i] = low[i] * (1 - copysign(shift_error, low[i])) - shift;
+          high[i] = high[i] * (1 + copysign(shift_error, high[i])) - shift;
+          eigenval_bisect_refine(l2, d2, low[i], high[i], i);
+          l_ptr = &l2;
+          d_ptr = &d2;
+        } else {
+          l_ptr = &block.l;
+          d_ptr = &block.d;
+        }
+        twist_idx = get_twisted_factorization(
+            *l_ptr, *d_ptr, (low[i] + high[i]) * 0.5, l_plus, u_minus);
+        calculate_eigenvector(l_plus, u_minus, subdiagonal, i, twist_idx,
+                              eigenvectors);
+      }
+    }
+  }
+}
+
+/**
+ * Calculates eigenvalues and eigenvectors of a tridiagonal matrix T using MRRR
+ * algorithm. If a subdiagonal element is close to zero compared to neighbors on
+ * diagonal the problem can be split into smaller ones.
+ * @param diagonal Diagonal of of T.
+ * @param subdiagonal Subdiagonal of T.
+ * @param[out] eigenvalues Eigenvlues.
+ * @param[out] eigenvectors Eigenvectors.
+ * @param split_threshold Threshold for splitting the problem
+ */
+void tridiagonal_eigensolver(const Eigen::VectorXd& diagonal,
```

Comment:
If this is the actual function that's supposed to be called, please call the name of this file `tridiagonal_eigensolver.hpp`.